___

<a href='http://www.pieriandata.com'><img src='../Pierian_Data_Logo.png'/></a>
___
<center><em>Copyright by Pierian Data Inc.</em></center>
<center><em>For more information, visit us at <a href='http://www.pieriandata.com'>www.pieriandata.com</a></em></center>

# DataFrames

Throughout the course, most of our data exploration will be done with DataFrames. DataFrames are an extremely powerful tool and a natural extension of the Pandas Series. By definition all a DataFrame is:

**A Pandas DataFrame consists of multiple Pandas Series that share index values.**

## Imports

In [2]:
import numpy as np
import pandas as pd

### Creating a DataFrame from Python Objects

In [3]:
# help(pd.DataFrame)

In [4]:
# Make sure the seed is in the same cell as the random call
# https://stackoverflow.com/questions/21494489/what-does-numpy-random-seed0-do
np.random.seed(101)
mydata=np.random.randint(1,100,(4,3))

In [8]:
np.random.seed(101)
mydata=np.random.randint(0,100,(4,3))
mydata

array([[95, 11, 81],
       [70, 63, 87],
       [75,  9, 77],
       [40,  4, 63]])

In [19]:
myindex = ['CA','NY','AZ','TX']

In [20]:
mycolumns = ['Jan','Feb','Mar']

In [26]:
df=pd.DataFrame(mydata)
df

0   1   2
0  95  11  81
1  70  63  87
2  75   9  77
3  40   4  63

In [31]:
df=pd.DataFrame(mydata,myindex)
df

0   1   2
CA  95  11  81
NY  70  63  87
AZ  75   9  77
TX  40   4  63

In [29]:
df = pd.DataFrame(data=mydata,index=myindex,columns=mycolumns)
df 

Jan  Feb  Mar
CA   95   11   81
NY   70   63   87
AZ   75    9   77
TX   40    4   63

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, CA to TX
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   0       4 non-null      int64
 1   1       4 non-null      int64
 2   2       4 non-null      int64
dtypes: int64(3)
memory usage: 128.0+ bytes



# Reading a .csv file for a DataFrame

----

## NOTE: We will go over all kinds of data inputs and outputs (.html, .csv, .xlxs , etc...) later on in the course! For now we just need to read in a simple .csv file.

----

## CSV
Comma Separated Values files are text files that use commas as field delimeters.<br>
Unless you're running the virtual environment included with the course, you may need to install <tt>xlrd</tt> and <tt>openpyxl</tt>.<br>
In your terminal/command prompt run:

    conda install xlrd
    conda install openpyxl

Then restart Jupyter Notebook.
(or use pip install if you aren't using the Anaconda Distribution)

### Understanding File Paths

You have two options when reading a file with pandas:

1. If your .py file or .ipynb notebook is located in the **exact** same folder location as the .csv file you want to read, simply pass in the file name as a string, for example:
    
        df = pd.read_csv('some_file.csv')
        
2. Pass in the entire file path if you are located in a different directory. The file path must be 100% correct in order for this to work. For example:

        df = pd.read_csv("C:\\Users\\myself\\files\\some_file.csv")

#### Print your current directory file path with pwd

In [33]:
pwd

'/Users/zhoulili/Documents/Udemy_Python/UNZIP_FOR_NOTEBOOKS_FINAL/03-Pandas'

#### List the files in your current directory with ls

In [34]:
ls

00-Series.ipynb
01-DataFrames.ipynb
02-Conditional-Filtering.ipynb
03-Useful-Methods.ipynb
04-Missing-Data.ipynb
05-Groupby-Operations-and-MultiIndex.ipynb
06-Combining-DataFrames.ipynb
07-Text-Methods.ipynb
08-Time-Methods.ipynb
09-Inputs-and-Outputs.ipynb
10-Pivot-Tables.ipynb
11-Pandas-Project-Exercise .ipynb
12-Pandas-Project-Exercise-Solution.ipynb
Copy1_00-Series.ipynb
Copy1_01-DataFrames.ipynb
RetailSales_BeerWineLiquor.csv
Sales_Funnel_CRM.csv
__pycache__/
example.csv
example.xlsx
hotel_booking_data.csv
movie_scores.csv
mpg.csv
my_excel_file.xlsx
new_file.csv
new_workbook.xlsx
newfile.csv
reshaping_pivot.png
sample_table.html
simple.html
tips.csv


In [35]:
df=pd.read_csv('tips.csv')

In [36]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

----
About this DataSet (in case you are interested)

* Description
    * One waiter recorded information about each tip he received over a period of a few months working in one restaurant. He collected several variables:

* Format
    * A data frame with 244 rows and 7 variables

* Details
    * tip in dollars,
    * bill in dollars,
    * sex of the bill payer,
    * whether there were smokers in the party,
    * day of the week,
    * time of day,
    * size of the party.

In all he recorded 244 tips. The data was reported in a collection of case studies for business statistics (Bryant & Smith 1995).

* References
    * Bryant, P. G. and Smith, M (1995) Practical Data Analysis: Case Studies in Business Statistics. Homewood, IL: Richard D. Irwin Publishing:
    
* Note: We created some additional columns with Fake data, including Name, CC Number, and Payment ID.

----

# DataFrames

## Obtaining Basic Information About DataFrame

In [38]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size',
       'price_per_person', 'Payer Name', 'CC Number', 'Payment ID'],
      dtype='object')

In [41]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

## df.head(-10) 负数就是删除后面的10行 输出删除后的database

In [43]:
df

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [42]:
df.head(-3)

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608

index  total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                               
Sat3880       241       22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17         242       17.82  1.75    Male     No   Sat  Dinner     2   
Thur672       243       18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person       Payer Name         CC Number  
Payment ID                                                       
Sat3880                11.34       Keith Wong  6011891618747196  
Sat17                   8.91     Dennis Dixon     4375220550950  
Thur672                 9.39  Michelle Hardin  3511451626698139

In [48]:
df.set_index('Payment ID')


total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

In [49]:
df

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [46]:
df.reset_index()

index  total_bill   tip     sex smoker  day    time  size  \
0      0       16.99  1.01  Female     No  Sun  Dinner     2   
1      1       10.34  1.66    Male     No  Sun  Dinner     3   
2      2       21.01  3.50    Male     No  Sun  Dinner     3   
3      3       23.68  3.31    Male     No  Sun  Dinner     2   
4      4       24.59  3.61  Female     No  Sun  Dinner     4   

   price_per_person          Payer Name         CC Number Payment ID  
0              8.49  Christy Cunningham  3560325168603410    Sun2959  
1              3.45      Douglas Tucker  4478071379779230    Sun4608  
2              7.00      Travis Walters  6011812112971322    Sun4458  
3             11.84    Nathaniel Harris  4676137647685994    Sun5260  
4              6.15        Tonya Carter  4832732618637221    Sun2251

In [47]:
df

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [71]:
len(df)

243

In [50]:
df.describe()

total_bill       tip     size  price_per_person     CC Number
count    5.000000  5.000000  5.00000          5.000000  5.000000e+00
mean    19.322000  2.618000  2.80000          7.386000  4.711816e+15
std      5.824686  1.198361  0.83666          3.091348  8.784150e+14
min     10.340000  1.010000  2.00000          3.450000  3.560325e+15
25%     16.990000  1.660000  2.00000          6.150000  4.478071e+15
50%     21.010000  3.310000  3.00000          7.000000  4.676138e+15
75%     23.680000  3.500000  3.00000          8.490000  4.832733e+15
max     24.590000  3.610000  4.00000         11.840000  6.011812e+15

In [51]:
df.describe().transpose()

count          mean           std           min  \
total_bill          5.0  1.932200e+01  5.824686e+00  1.034000e+01   
tip                 5.0  2.618000e+00  1.198361e+00  1.010000e+00   
size                5.0  2.800000e+00  8.366600e-01  2.000000e+00   
price_per_person    5.0  7.386000e+00  3.091348e+00  3.450000e+00   
CC Number           5.0  4.711816e+15  8.784150e+14  3.560325e+15   

                           25%           50%           75%           max  
total_bill        1.699000e+01  2.101000e+01  2.368000e+01  2.459000e+01  
tip               1.660000e+00  3.310000e+00  3.500000e+00  3.610000e+00  
size              2.000000e+00  3.000000e+00  3.000000e+00  4.000000e+00  
price_per_person  6.150000e+00  7.000000e+00  8.490000e+00  1.184000e+01  
CC Number         4.478071e+15  4.676138e+15  4.832733e+15  6.011812e+15

## Selection and Indexing

Let's learn how to retrieve information from a DataFrame.

### COLUMNS

We will begin be learning how to extract information based on the columns

In [68]:
df

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

#### Grab a Single Column

In [69]:
df[['total_bill']]

total_bill
Payment ID            
Sun2959          16.99
Sun4608          10.34
Sun4458          21.01
Sun5260          23.68
Sun2251          24.59

In [70]:
df[['total_bill','tip']]

total_bill   tip
Payment ID                  
Sun2959          16.99  1.01
Sun4608          10.34  1.66
Sun4458          21.01  3.50
Sun5260          23.68  3.31
Sun2251          24.59  3.61

In [74]:
df.iloc[0]

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                  8.49
Payer Name          Christy Cunningham
CC Number             3560325168603410
Name: Sun2959, dtype: object

In [76]:
df.loc[['Sun2959','Sun4608']]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230

KeyError: 0

In [79]:
df.read()

AttributeError: 'DataFrame' object has no attribute 'read'

#### Grab Multiple Columns

In [84]:
# Note how its a python list of column names! Thus the double brackets.


total_bill   tip
0         16.99  1.01
1         10.34  1.66
2         21.01  3.50
3         23.68  3.31
4         24.59  3.61
..          ...   ...
239       29.03  5.92
240       27.18  2.00
241       22.67  2.00
242       17.82  1.75
243       18.78  3.00

[244 rows x 2 columns]

#### Create New Columns

In [86]:
df['tips_per']=df['tip']/df['total_bill']*100

In [87]:
df

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2              8.49   
1         10.34  1.66    Male     No   Sun  Dinner     3              3.45   
2         21.01  3.50    Male     No   Sun  Dinner     3              7.00   
3         23.68  3.31    Male     No   Sun  Dinner     2             11.84   
4         24.59  3.61  Female     No   Sun  Dinner     4              6.15   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3              9.68   
240       27.18  2.00  Female    Yes   Sat  Dinner     2             13.59   
241       22.67  2.00    Male    Yes   Sat  Dinner     2             11.34   
242       17.82  1.75    Male     No   Sat  Dinner     2              8.91   
243       18.78  3.00  Female     No  Thur  Dinner     2              9.39   

             Payer Name         CC Number Payment ID   tips_per  
0    Christy Cunningham  3560325168603410    Sun2959   5.944673  
1        Douglas Tucker  4478071379779230    Sun4608  16.054159  
2        Travis Walters  6011812112971322    Sun4458  16.658734  
3      Nathaniel Harris  4676137647685994    Sun5260  13.978041  
4          Tonya Carter  4832732618637221    Sun2251  14.680765  
..                  ...               ...        ...        ...  
239       Michael Avila  5296068606052842    Sat2657  20.392697  
240      Monica Sanders  3506806155565404    Sat1766   7.358352  
241          Keith Wong  6011891618747196    Sat3880   8.822232  
242        Dennis Dixon     4375220550950      Sat17   9.820426  
243     Michelle Hardin  3511451626698139    Thur672  15.974441  

[244 rows x 12 columns]

In [88]:
df['price_per_person']=df['total_bill']/df['size']

In [89]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.66    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.50    Male     No  Sun  Dinner     3          7.003333   
3       23.68  3.31    Male     No  Sun  Dinner     2         11.840000   
4       24.59  3.61  Female     No  Sun  Dinner     4          6.147500   

           Payer Name         CC Number Payment ID   tips_per  
0  Christy Cunningham  3560325168603410    Sun2959   5.944673  
1      Douglas Tucker  4478071379779230    Sun4608  16.054159  
2      Travis Walters  6011812112971322    Sun4458  16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260  13.978041  
4        Tonya Carter  4832732618637221    Sun2251  14.680765

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.66    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.50    Male     No  Sun  Dinner     3          7.003333   
3       23.68  3.31    Male     No  Sun  Dinner     2         11.840000   
4       24.59  3.61  Female     No  Sun  Dinner     4          6.147500   

           Payer Name         CC Number Payment ID  tip_percentage  
0  Christy Cunningham  3560325168603410    Sun2959        5.944673  
1      Douglas Tucker  4478071379779230    Sun4608       16.054159  
2      Travis Walters  6011812112971322    Sun4458       16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260       13.978041  
4        Tonya Carter  4832732618637221    Sun2251       14.680765

In [78]:
help(np.round)

Help on function round_ in module numpy:

round_(a, decimals=0, out=None)
    Round an array to the given number of decimals.
    
    See Also
    --------
    around : equivalent function; see for details.



#### Adjust Existing Columns

In [79]:
# Because pandas is based on numpy, we get awesome capabilities with numpy's universal functions!


In [90]:
df.head()


total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2          8.495000   
1       10.34  1.66    Male     No  Sun  Dinner     3          3.446667   
2       21.01  3.50    Male     No  Sun  Dinner     3          7.003333   
3       23.68  3.31    Male     No  Sun  Dinner     2         11.840000   
4       24.59  3.61  Female     No  Sun  Dinner     4          6.147500   

           Payer Name         CC Number Payment ID   tips_per  
0  Christy Cunningham  3560325168603410    Sun2959   5.944673  
1      Douglas Tucker  4478071379779230    Sun4608  16.054159  
2      Travis Walters  6011812112971322    Sun4458  16.658734  
3    Nathaniel Harris  4676137647685994    Sun5260  13.978041  
4        Tonya Carter  4832732618637221    Sun2251  14.680765

#### Remove Columns

In [91]:
# df.drop('tip_percentage',axis=1)
df.drop('tips_per',axis=1)

total_bill   tip     sex smoker   day    time  size  price_per_person  \
0         16.99  1.01  Female     No   Sun  Dinner     2          8.495000   
1         10.34  1.66    Male     No   Sun  Dinner     3          3.446667   
2         21.01  3.50    Male     No   Sun  Dinner     3          7.003333   
3         23.68  3.31    Male     No   Sun  Dinner     2         11.840000   
4         24.59  3.61  Female     No   Sun  Dinner     4          6.147500   
..          ...   ...     ...    ...   ...     ...   ...               ...   
239       29.03  5.92    Male     No   Sat  Dinner     3          9.676667   
240       27.18  2.00  Female    Yes   Sat  Dinner     2         13.590000   
241       22.67  2.00    Male    Yes   Sat  Dinner     2         11.335000   
242       17.82  1.75    Male     No   Sat  Dinner     2          8.910000   
243       18.78  3.00  Female     No  Thur  Dinner     2          9.390000   

             Payer Name         CC Number Payment ID  
0    Christy Cunningham  3560325168603410    Sun2959  
1        Douglas Tucker  4478071379779230    Sun4608  
2        Travis Walters  6011812112971322    Sun4458  
3      Nathaniel Harris  4676137647685994    Sun5260  
4          Tonya Carter  4832732618637221    Sun2251  
..                  ...               ...        ...  
239       Michael Avila  5296068606052842    Sat2657  
240      Monica Sanders  3506806155565404    Sat1766  
241          Keith Wong  6011891618747196    Sat3880  
242        Dennis Dixon     4375220550950      Sat17  
243     Michelle Hardin  3511451626698139    Thur672  

[244 rows x 11 columns]

In [84]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

# Index Basics

Before going over the same retrieval tasks for rows, let's build some basic understanding of the pandas DataFrame Index.

In [85]:
df.head()

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [92]:
df.index

RangeIndex(start=0, stop=244, step=1)

In [96]:
df=df.set_index('Payment ID')

total_bill   tip     sex smoker  day    time  size  price_per_person  \
0       16.99  1.01  Female     No  Sun  Dinner     2              8.49   
1       10.34  1.66    Male     No  Sun  Dinner     3              3.45   
2       21.01  3.50    Male     No  Sun  Dinner     3              7.00   
3       23.68  3.31    Male     No  Sun  Dinner     2             11.84   
4       24.59  3.61  Female     No  Sun  Dinner     4              6.15   

           Payer Name         CC Number Payment ID  
0  Christy Cunningham  3560325168603410    Sun2959  
1      Douglas Tucker  4478071379779230    Sun4608  
2      Travis Walters  6011812112971322    Sun4458  
3    Nathaniel Harris  4676137647685994    Sun5260  
4        Tonya Carter  4832732618637221    Sun2251

In [97]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number   tips_per  
Payment ID                                                                     
Sun2959             8.495000  Christy Cunningham  3560325168603410   5.944673  
Sun4608             3.446667      Douglas Tucker  4478071379779230  16.054159  
Sun4458             7.003333      Travis Walters  6011812112971322  16.658734  
Sun5260            11.840000    Nathaniel Harris  4676137647685994  13.978041  
Sun2251             6.147500        Tonya Carter  4832732618637221  14.680765

In [98]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number   tips_per  
Payment ID                                                                     
Sun2959             8.495000  Christy Cunningham  3560325168603410   5.944673  
Sun4608             3.446667      Douglas Tucker  4478071379779230  16.054159  
Sun4458             7.003333      Travis Walters  6011812112971322  16.658734  
Sun5260            11.840000    Nathaniel Harris  4676137647685994  13.978041  
Sun2251             6.147500        Tonya Carter  4832732618637221  14.680765

### ROWS

Let's now explore these same concepts but with Rows.

In [95]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   

            price_per_person          Payer Name         CC Number  
Payment ID                                                          
Sun2959                 8.49  Christy Cunningham  3560325168603410  
Sun4608                 3.45      Douglas Tucker  4478071379779230  
Sun4458                 7.00      Travis Walters  6011812112971322  
Sun5260                11.84    Nathaniel Harris  4676137647685994  
Sun2251                 6.15        Tonya Carter  4832732618637221

#### Grab a Single Row

In [99]:
# Integer Based
df.iloc[0]

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                 8.495
Payer Name          Christy Cunningham
CC Number             3560325168603410
tips_per                       5.94467
Name: Sun2959, dtype: object

In [100]:
# Name Based
df.loc['Sun2959']

total_bill                       16.99
tip                               1.01
sex                             Female
smoker                              No
day                                Sun
time                            Dinner
size                                 2
price_per_person                 8.495
Payer Name          Christy Cunningham
CC Number             3560325168603410
tips_per                       5.94467
Name: Sun2959, dtype: object

#### Grab Multiple Rows

In [101]:
df.iloc[0:2]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   

            price_per_person          Payer Name         CC Number   tips_per  
Payment ID                                                                     
Sun2959             8.495000  Christy Cunningham  3560325168603410   5.944673  
Sun4608             3.446667      Douglas Tucker  4478071379779230  16.054159

In [102]:
df.loc[['Sun2959','Sun5260']]

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun2959          16.99  1.01  Female     No  Sun  Dinner     2   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   

            price_per_person          Payer Name         CC Number   tips_per  
Payment ID                                                                     
Sun2959                8.495  Christy Cunningham  3560325168603410   5.944673  
Sun5260               11.840    Nathaniel Harris  4676137647685994  13.978041

#### Remove Row

Typically are datasets will be large enough that we won't remove rows like this since we won't know thier row location for some specific condition, instead, we drop rows based on conditions such as missing data or column values. The next lecture will cover this in a lot more detail.

In [106]:
df=df.drop('Sun2959',axis=0)

In [107]:
df.head()

total_bill   tip     sex smoker  day    time  size  \
Payment ID                                                       
Sun4608          10.34  1.66    Male     No  Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No  Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No  Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No  Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No  Sun  Dinner     4   

            price_per_person        Payer Name         CC Number   tips_per  
Payment ID                                                                   
Sun4608             3.446667    Douglas Tucker  4478071379779230  16.054159  
Sun4458             7.003333    Travis Walters  6011812112971322  16.658734  
Sun5260            11.840000  Nathaniel Harris  4676137647685994  13.978041  
Sun2251             6.147500      Tonya Carter  4832732618637221  14.680765  
Sun9679             6.322500        Erik Smith   213140353657882  18.623962

In [102]:
# Error if you have a named index!
# df.drop(0,axis=0).head()

#### Insert a New Row

Pretty rare to add a single row like this. Usually you use pd.concat() to add many rows at once. You could use the .append() method with a list of pd.Series() objects, but you won't see us do this with realistic real-world data.

In [108]:
one_row=df.iloc[0]

In [109]:
one_row

total_bill                     10.34
tip                             1.66
sex                             Male
smoker                            No
day                              Sun
time                          Dinner
size                               3
price_per_person             3.44667
Payer Name            Douglas Tucker
CC Number           4478071379779230
tips_per                     16.0542
Name: Sun4608, dtype: object

In [110]:

type(one_row)

pandas.core.series.Series

In [111]:
df.tail()

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person       Payer Name         CC Number   tips_per  
Payment ID                                                                  
Sat2657             9.676667    Michael Avila  5296068606052842  20.392697  
Sat1766            13.590000   Monica Sanders  3506806155565404   7.358352  
Sat3880            11.335000       Keith Wong  6011891618747196   8.822232  
Sat17               8.910000     Dennis Dixon     4375220550950   9.820426  
Thur672             9.390000  Michelle Hardin  3511451626698139  15.974441

In [113]:
# df.append(one_row).tail()
df.append(one_row).tail(1)

total_bill   tip   sex smoker  day    time  size  \
Payment ID                                                     
Sun4608          10.34  1.66  Male     No  Sun  Dinner     3   

            price_per_person      Payer Name         CC Number   tips_per  
Payment ID                                                                 
Sun4608             3.446667  Douglas Tucker  4478071379779230  16.054159

In [118]:
df['price_per_person']=np.round(df['price_per_person'],2)
df

total_bill   tip     sex smoker   day    time  size  \
Payment ID                                                        
Sun4608          10.34  1.66    Male     No   Sun  Dinner     3   
Sun4458          21.01  3.50    Male     No   Sun  Dinner     3   
Sun5260          23.68  3.31    Male     No   Sun  Dinner     2   
Sun2251          24.59  3.61  Female     No   Sun  Dinner     4   
Sun9679          25.29  4.71    Male     No   Sun  Dinner     4   
...                ...   ...     ...    ...   ...     ...   ...   
Sat2657          29.03  5.92    Male     No   Sat  Dinner     3   
Sat1766          27.18  2.00  Female    Yes   Sat  Dinner     2   
Sat3880          22.67  2.00    Male    Yes   Sat  Dinner     2   
Sat17            17.82  1.75    Male     No   Sat  Dinner     2   
Thur672          18.78  3.00  Female     No  Thur  Dinner     2   

            price_per_person        Payer Name         CC Number   tips_per  
Payment ID                                                                   
Sun4608                 3.45    Douglas Tucker  4478071379779230  16.054159  
Sun4458                 7.00    Travis Walters  6011812112971322  16.658734  
Sun5260                11.84  Nathaniel Harris  4676137647685994  13.978041  
Sun2251                 6.15      Tonya Carter  4832732618637221  14.680765  
Sun9679                 6.32        Erik Smith   213140353657882  18.623962  
...                      ...               ...               ...        ...  
Sat2657                 9.68     Michael Avila  5296068606052842  20.392697  
Sat1766                13.59    Monica Sanders  3506806155565404   7.358352  
Sat3880                11.34        Keith Wong  6011891618747196   8.822232  
Sat17                   8.91      Dennis Dixon     4375220550950   9.820426  
Thur672                 9.39   Michelle Hardin  3511451626698139  15.974441  

[243 rows x 11 columns]

--------